### ISSUE TITLE
WebAuthRequired from TransferItem

### ISSUE THREAD

**Issue by MarkAlmighty** [Bungie-net/api#1201]
As the subject suggests, I'm getting WebAuthRequired back from my attempts to TransferItem with PHP/CURL. I've scoured these forums to no avail. Maybe I'm missing a curl_setopt() ? I'm cutting my API teeth on this project, so this is all new to me.

- I have "www" in the endpoint URL.
- I have a "/" on the end of the endpoint URL.
- I have created a new API key, updated my OAuth credentials, and I definitely have an access token that has not expired.

Here is my code:

```
$curl = curl_init('https://www.bungie.net/Platform/Destiny2/Actions/Items/TransferItem/');

curl_setopt($curl, CURLOPT_HEADER, false);
curl_setopt($curl, CURLOPT_RETURNTRANSFER, true);

// POST *
curl_setopt($curl, CURLOPT_POSTFIELDS, json_encode(array(
	'itemReferenceHash' => $itemHash,         // uint32
	'stackSize' => 1,                                       // int32
	'transferToVault' => false,                         // boolean
	'itemId' => $itemId,                                 // int64
	'characterId' => $character_id,                 // int64
	'membershipType' => 3                           // int32
)));

// HTTPHEADER *
curl_setopt($curl, CURLOPT_HTTPHEADER, json_encode(array(
  "Authorization" => "Bearer $access_token",
  "Content-Type" => "application/json"
)));

$response = curl_exec($curl);

curl_close($curl);
```

**Comment by Tetron-bng** [comment]
I am not a PHP, curl expert, but at least some samples I found on Stack Overflow didn't json_encode the array. It looked more like this:

$headers = array(
    'Authorization: Bearer $access_token",
  "Content-Type: application/json"
);
curl_setopt($curl, CURLOPT_HTTPHEADER, $headers);

**Comment by justrealmilk** [comment]
so this is how I did back in the day

```php
    $defaultCurlOptions = array(
      CURLOPT_USERAGENT => USER_AGENT,
      CURLOPT_RETURNTRANSFER => true,
      CURLOPT_FOLLOWLOCATION => true,
      CURLOPT_SSL_VERIFYHOST => 2,
      CURLOPT_HTTPHEADER => array(
        'X-API-Key: ' . API_KEY
      )
    );

    $ch = curl_init();
    curl_setopt_array($ch, $defaultCurlOptions);
    curl_setopt($ch, CURLOPT_URL,'https://www.bungie.net/Platform/Destiny2/1/Profile/46116869000000000000/Character/2305800000000000004/Vendors/' . $vendorHashValue . '/?components=400,401,402');
    curl_setopt($ch, CURLOPT_HTTPHEADER, array(
      'Authorization: Bearer ' . $tokens->access_token
    ));

    $response = curl_exec($ch);
    $response = json_decode($response);
    curl_close($ch);

    if ($response->ErrorCode == 1) {
```

**Comment by MarkAlmighty** [comment]
> 
> 
> I am not a PHP, curl expert, but at least some samples I found on Stack Overflow didn't json_encode the array. It looked more like this:
> 
> $headers = array(
> 'Authorization: Bearer $access_token",
> "Content-Type: application/json"
> );
> curl_setopt($curl, CURLOPT_HTTPHEADER, $headers);

I removed json_encode from the arrays in both CURLOPT_POSTFIELDS and CURLOPT_HTTPHEADER, but I still get back ErrorStatus => WebAuthRequired.

**Comment by MarkAlmighty** [comment]
> 
> 
> so this is how I did back in the day
> 
> ```
>     $defaultCurlOptions = array(
>       CURLOPT_USERAGENT => USER_AGENT,
>       CURLOPT_RETURNTRANSFER => true,
>       CURLOPT_FOLLOWLOCATION => true,
>       CURLOPT_SSL_VERIFYHOST => 2,
>       CURLOPT_HTTPHEADER => array(
>         'X-API-Key: ' . API_KEY
>       )
>     );
> 
>     $ch = curl_init();
>     curl_setopt_array($ch, $defaultCurlOptions);
>     curl_setopt($ch, CURLOPT_URL,'https://www.bungie.net/Platform/Destiny2/1/Profile/46116869000000000000/Character/2305800000000000004/Vendors/' . $vendorHashValue . '/?components=400,401,402');
>     curl_setopt($ch, CURLOPT_HTTPHEADER, array(
>       'Authorization: Bearer ' . $tokens->access_token
>     ));
> 
>     $response = curl_exec($ch);
>     $response = json_decode($response);
>     curl_close($ch);
> 
>     if ($response->ErrorCode == 1) {
> ```

I haven't found anything from in your example that has changed the outcome, but thank you.

**Comment by MarkAlmighty** [comment]
> 
> 
> Do other requests work with the access token you have, or is the only one you have tried? Do you have any tools that let you see exactly what the request you send looks like? For example on the PC, Fiddler is very useful. Postman is good on Mac. Compare what you are sending to what something like Destiny Item Manager sends (via Chrome debug tools).
> 
> Often the problem is the code is not producing the request you think it producing. Starting there we can work back to the code and figure out why the format is incorrect.
> 
> OH! Also make sure you include your API key in the list of headers.

This is my first attempt (ever) to access an endpoint that requires a Bearer access token. I've tried setting this up in Postman, but I'm doing something even more wrong there as I keep getting back a 500 Internal Server Error. I'll see if I can figure out how to use browser debug tools with DIM like you've suggested.

Since I'm self-teaching API usage with this D2 app, I know there are a lot of API basics that I just don't know yet. I've learned a ton so far, and I've been able to power through figuring out the GET endpoints, but this POST is stumping me pretty hard. When I try to Google for solutions, I'm only finding bits and pieces, and usually about other endpoints.

I'm kind of just throwing mud at the wall to see what sticks. Like:
- Should I be json_encode-ing the CURLOPT_POSTFIELDS and CURLOPT_HTTPHEADER arrays?
- Should the Content-type be 'application/json' or 'application/x-www-form-urlencoded'?
- Are there certain curl_setopt's that I'm missing?

I haven't found where stuff like this is spelled out from beginning to end, and perhaps I shouldn't expect it to be. I'm not really complaining (or I'm trying not to), I'm just flailing. :)

**Comment by MarkAlmighty** [comment]
Oh, and I've tried it both with and without my X-API-KEY.

**Comment by justrealmilk** [comment]
You'll always need your API key.

Considering your first two bullet points from your last comment, I think they're both required and that they always need match. So if you send a JSON_ENCODED payload, specify that it's JSON. I'm sure you know this and that it hasn't helped!

Maybe there's something hinky with your application config over at bnet?

As for spying on how other tools do it, you can try this:

![image](https://user-images.githubusercontent.com/156681/76050668-015f5b80-5fb4-11ea-9bc1-85a9d0e3eaaa.png)

**Comment by justrealmilk** [comment]
This is me transferring a weapon from character to vault with DIM:

```
curl 'https://www.bungie.net/Platform/Destiny2/Actions/Items/TransferItem/' \
  -H 'authority: www.bungie.net' \
  -H 'pragma: no-cache' \
  -H 'cache-control: no-cache' \
  -H 'authorization: Bearer SecretTomBusiness' \
  -H 'x-api-key: 2b16c291fcff48cbac86bd0f1d0bbc90' \
  -H 'content-type: application/json' \
  -H 'origin: https://beta.destinyitemmanager.com' \
  --data-binary '{"characterId":"2305843009359344999","membershipType":1,"itemId":"6917529145907333045","itemReferenceHash":2069224589,"stackSize":1,"transferToVault":true}' \
  --compressed
```

So there's the content type and encoded payload. If you haven't already, you could try verifying that your payload is sound by printing it out

**Comment by Tetron-bng** [comment]
Doing this type of work without the use of tools (Postman, Fiddler, Chrome Debugger) is a bit like driving a car with just the instrument panel.  You can push the peddles and turn the wheel, but it's really hard to know what you are running into.  I recommend you take the time to figure out the tools. It is amazing how much more clear things will become.

**Comment by MarkAlmighty** [comment]
You're both giving me good stuff to look at. I'm checking out Fiddler now. I'll take a step back and put some time in to learning these tools better. If I get this working and come back here and lay out the solution.

**Comment by MarkAlmighty** [comment]
Well, I figured it out. And I'm equal parts excited and embarrassed. 

So by using Postman, I finally realized that it was the 500 Internal Server Errors I should have been chasing instead of the WebAuthRequired. It dawned on me when I excluded the Authorization piece and got the 500. I had been seeing them during some of this troubleshooting, but never in a context that helped. Anyways, going on the assumption there was something wrong with my token (and I knew it wasn't expired), I started looking at all of the structure involved. I eventually figured out that I was storing both tokens in a database column with data type of TINYTEXT. Which has a maximum length of 255 characters. And the tokens are 360. /headdesk

Once I switched to the TEXT data type and got a new token, I was able to move my Jade Rabbit from my Vault to my Warlock. And **_THAT_** was some exhilarating shiznit!

Thanks again @justrealmilk and @Tetron-bng for your responses!
